### ISSUE TITLE
Planner crash: invalid attrno

### ISSUE THREAD

**Issue by oSkrobuk** [greenplum-db/gpdb#10720]
### Greenplum version or build
PostgreSQL 9.4.24 (Greenplum Database 6.9.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609, 64-bit compiled on Aug  5 2020 20:56:00

### OS version and uname -a
Linux hostname 4.19.62-13 #1 SMP Mon Jul 29 17:22:59 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### Installation information ( pg_config )
BINDIR = /usr/local/gpdb/bin
DOCDIR = /usr/local/gpdb/share/doc/postgresql
HTMLDIR = /usr/local/gpdb/share/doc/postgresql
INCLUDEDIR = /usr/local/gpdb/include
PKGINCLUDEDIR = /usr/local/gpdb/include/postgresql
INCLUDEDIR-SERVER = /usr/local/gpdb/include/postgresql/server
LIBDIR = /usr/local/gpdb/lib/x86_64-linux-gnu
PKGLIBDIR = /usr/local/gpdb/lib/x86_64-linux-gnu/postgresql
LOCALEDIR = /usr/local/gpdb/share/locale
MANDIR = /usr/local/gpdb/share/man
SHAREDIR = /usr/local/gpdb/share/postgresql
SYSCONFDIR = /etc/postgresql
PGXS = /usr/local/gpdb/lib/x86_64-linux-gnu/postgresql/pgxs/src/makefiles/pgxs.mk
CONFIGURE = '--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=/usr/local/gpdb/include' '--mandir=/usr/local/gpdb/share/man' '--infodir=/usr/local/gpdb/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=/usr/local/gpdb/lib/x86_64-linux-gnu' '--libexecdir=/usr/local/gpdb/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--with-perl' '--with-python' '--without-gssapi' '--disable-gpcloud' '--with-libxml' '--with-system-tzdata=/usr/share/zoneinfo' '--prefix=/usr/local/gpdb' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security'
CC = gcc
CPPFLAGS = -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2 -I/usr/local/gpdb/include
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -std=gnu99 -Werror=uninitialized -Werror=implicit-function-declaration -I/usr/local/gpdb/include
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,--as-needed -Wl,-rpath,'/usr/local/gpdb/lib/x86_64-linux-gnu',--enable-new-dtags -L/usr/local/gpdb/lib/x86_64-linux-gnu
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -lxerces-c -lxml2 -lrt -lzstd -lrt -lcrypt -ldl -lm -L/usr/local/gpdb/lib/x86_64-linux-gnu
VERSION = PostgreSQL 9.4.24

### Actual behavior
`invalid attrno 9, should be in the range [1, 8] (cdbgroup.c:4436)`
```
Stack trace:
1    0xae0ee8 postgres errstart + 0x248
2    0xae3aa0 postgres elog_finish + 0xd0
3    0xb62891 postgres <symbol not found> + 0xb62891
4    0x8da643 postgres expression_tree_mutator + 0x333
5    0x8da6b4 postgres expression_tree_mutator + 0x3a4
6    0x8da643 postgres expression_tree_mutator + 0x333
7    0x8da6b4 postgres expression_tree_mutator + 0x3a4
8    0x8da643 postgres expression_tree_mutator + 0x333
9    0x8da6b4 postgres expression_tree_mutator + 0x3a4
10   0x8da643 postgres expression_tree_mutator + 0x333
11   0x8da6b4 postgres expression_tree_mutator + 0x3a4
12   0x8da643 postgres expression_tree_mutator + 0x333
13   0x8da6b4 postgres expression_tree_mutator + 0x3a4
14   0x8da643 postgres expression_tree_mutator + 0x333
15   0x8da6b4 postgres expression_tree_mutator + 0x3a4
16   0x8da7dc postgres expression_tree_mutator + 0x4cc
17   0x8da643 postgres expression_tree_mutator + 0x333
18   0x8da3de postgres expression_tree_mutator + 0xce
19   0x8da6ff postgres expression_tree_mutator + 0x3ef
20   0x8da643 postgres expression_tree_mutator + 0x333
21   0xb629b3 postgres <symbol not found> + 0xb629b3
22   0xb64b29 postgres <symbol not found> + 0xb64b29
23   0xb6832a postgres cdb_grouping_planner + 0x244a
24   0x940d0b postgres <symbol not found> + 0x940d0b
25   0x9424cd postgres subquery_planner + 0x76d
26   0x9428b1 postgres standard_planner + 0x181
27   0x943075 postgres planner + 0x165
28   0x9d8b0c postgres pg_plan_query + 0x1c
29   0x9d8c04 postgres pg_plan_queries + 0x54
30   0x9d9156 postgres <symbol not found> + 0x9d9156
```

### Step to reproduce the behavior
See next comment

**Comment by oSkrobuk** [comment]
The bug remains in 6.10.1

**Comment by oSkrobuk** [comment]
Some more examples. Errors:
```(sql)
--ERROR: invalid attrno 2, should be in the range [1, 1] (cdbgroup.c:4436)
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (
  id      VARCHAR,
  dttm    TIMESTAMP WITHOUT TIME ZONE
);
SELECT ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
  , CASE WHEN NOT
       ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
        THEN 'Previous'
    END
FROM foo t0
GROUP BY 1;
```
Successfully:
```(sql)
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (
  dttm    TIMESTAMP WITHOUT TIME ZONE,
  id      VARCHAR
);
SELECT ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
  , CASE WHEN NOT
       ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
        THEN 'Previous'
    END
FROM foo t0
GROUP BY 1;
```
`GROUP BY 1` does not the first column in the `SELECT`, but the first column in the table.

**Comment by kainwen** [comment]
Do not spend much time on this. It is a bug when creating two-stage agg plan. This part of code is re-wrote in master branch.

To work around:
1. try orca
2. under planner, force use one stage agg (might losing some perfmance under some condition):
    `set gp_eager_one_phase_agg = on;  set gp_eager_two_phase_agg = off;`

To fix this need more time to look into it.

**Comment by oSkrobuk** [comment]
> Do not spend much time on this. It is a bug when creating two-stage agg plan. This part of code is re-wrote in master branch.
> 
> To work around:
> 
> 1. try orca
> 2. under planner, force use one stage agg (might losing some perfmance under some condition):
>    `set gp_eager_one_phase_agg = on;  set gp_eager_two_phase_agg = off;`
> 
> To fix this need more time to look into it.
1. Does not work with ORCA
2. It worked

Thank

**Comment by oSkrobuk** [comment]
> 2. under planner, force use one stage agg (might losing some perfmance under some condition):
>    `set gp_eager_one_phase_agg = on;  set gp_eager_two_phase_agg = off;`

How to set gp_eager_one_phase_agg to=on by gpconfig? 
```
$ gpconfig -c gp_eager_one_phase_agg -v on
[CRITICAL]:-GUC Validation Failed: gp_eager_one_phase_agg cannot be changed under normal conditions. Please refer to gpconfig documentation.
GUC Validation Failed: gp_eager_one_phase_agg cannot be changed under normal conditions. Please refer to gpconfig documentation.
$ gpconfig -c gp_eager_one_phase_agg -v on --skipvalidation
20200902:20:11:56:062087 [INFO]:-completed successfully with parameters '-c gp_eager_one_phase_agg -v on --skipvalidation'
$ gpconfig -s gp_eager_one_phase_agg
Values on all segments are consistent
GUC          : gp_eager_one_phase_agg
Master  value: off
Segment value: off

```

**Comment by gaos1** [comment]
reopen this as only workaround works, but we need to check why two-stage agg error out.

**Comment by gfphoenix78** [comment]
@oSkrobuk 
```SQL
DROP TABLE IF EXISTS foo;
CREATE TABLE foo (
  dttm    TIMESTAMP WITHOUT TIME ZONE,
  id      VARCHAR
);
SELECT ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
  , CASE WHEN NOT
       ((DATE_TRUNC('DAY', t0.dttm::TIMESTAMP))::DATE >= (CURRENT_DATE + -365 * INTERVAL '1 DAY'))
        THEN 'Previous'
    END
FROM foo t0
GROUP BY 1;
```
The query doesn't raise an error, but it produces incorrect results. A simple example is:
```SQL
create table bar(a int, b int);

gpadmin=# insert into bar values(0, -1),(1, -2), (2,2), (2,3), (4,4);
INSERT 0 5
gpadmin=# select a>=2, not a>=2 from bar group by 1;
 ?column? | ?column?
----------+----------
 t        | t
 f        | t
```

**Comment by oSkrobuk** [comment]
@gfphoenix78 
If the first column from the table is selected for comparison and 1 is indicated in the GROUP BY, then it works without error. But it is worth choosing another column in comparison, and leaving 1 in GROUP BY, then get an error.
```sql
select b>=2, not b>=2 from bar group by 1;
```

**Comment by gfphoenix78** [comment]
> @gfphoenix78
> If the first column from the table is selected for comparison and 1 is indicated in the GROUP BY, then it works without error. But it is worth choosing another column in comparison, and leaving 1 in GROUP BY, then get an error.
> 
> ```sql
> select b>=2, not b>=2 from bar group by 1;
> --[XX000] ERROR: invalid attrno 2, should be in the range [1, 1] (cdbgroup.c:4436)
> ```

Yes, `select a>=2, not a>=2 from bar group by 1;` does just accidentally not raise an error. The root causes are the same. `not b >= 2` is converted to `b<2` internally, that's the point.

**Comment by kainwen** [comment]
Fixed by a8b33fc13010d44312e956ac797858396ac8b9f9
